{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "from datetime import datetime\n",
    "from datetime import timedelta\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from sklearn.datasets import fetch_20newsgroups\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in /opt/conda/lib/python3.7/site-packages (2.10.0)\n",
      "Requirement already satisfied: sentencepiece in /opt/conda/lib/python3.7/site-packages (from transformers) (0.1.91)\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.7/site-packages (from transformers) (1.18.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /opt/conda/lib/python3.7/site-packages (from transformers) (2019.8.19)\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.7/site-packages (from transformers) (3.0.10)\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.7/site-packages (from transformers) (2.23.0)\n",
      "Requirement already satisfied: tokenizers==0.7.0 in /opt/conda/lib/python3.7/site-packages (from transformers) (0.7.0)\n",
      "Requirement already satisfied: tqdm>=4.27 in /opt/conda/lib/python3.7/site-packages (from transformers) (4.45.0)\n",
      "Requirement already satisfied: sacremoses in /opt/conda/lib/python3.7/site-packages (from transformers) (0.0.43)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests->transformers) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests->transformers) (2020.4.5.1)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests->transformers) (1.24.3)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests->transformers) (2.9)\n",
      "Requirement already satisfied: joblib in /opt/conda/lib/python3.7/site-packages (from sacremoses->transformers) (0.14.1)\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.7/site-packages (from sacremoses->transformers) (1.14.0)\n",
      "Requirement already satisfied: click in /opt/conda/lib/python3.7/site-packages (from sacremoses->transformers) (7.1.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers\n",
    "import torch\n",
    "from transformers import BertForQuestionAnswering\n",
    "model = BertForQuestionAnswering.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')\n",
    "from transformers import BertTokenizer\n",
    "tokenizer = BertTokenizer.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_pdf_df = pd.read_csv(\"/home/jupyter/covid19-challenge/data/clean_doc_pdf.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1d7234df1a19a0cf0b4038ceb31e7a97881cdbd7</td>\n",
       "      <td>Virtual application of in situ simulation duri...</td>\n",
       "      <td>Erich Hanel, E M , Monika Bilic, Kelly Hassall...</td>\n",
       "      <td>Erich Hanel, E M , Monika Bilic, Kelly Hassall...</td>\n",
       "      <td>Abstract\\n\\nThe coronavirus disease 2019 (COVI...</td>\n",
       "      <td>\\n\\nRÉSUMÉ L'arrivée de la pandémie causée par...</td>\n",
       "      <td>In situ simulation and its effects on patient ...</td>\n",
       "      <td>[{'first': 'Erich', 'middle': [], 'last': 'Han...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'In situ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0ced1f946cce007aa319a0ba38aef2c4b14dab0e</td>\n",
       "      <td>BMC Immunology Identification of a novel conse...</td>\n",
       "      <td>Yanbo Lv, Zhihua Ruan, Li Wang, Bing Ni, Yuzha...</td>\n",
       "      <td>Yanbo Lv (Third Military Medical University, 3...</td>\n",
       "      <td>Abstract\\n\\nThe spike (S) protein is a major s...</td>\n",
       "      <td>Background\\n\\nSevere acute respiratory syndrom...</td>\n",
       "      <td>Identification of a novel coronavirus in patie...</td>\n",
       "      <td>[{'first': 'Yanbo', 'middle': [], 'last': 'Lv'...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2a667552f5924857d6eaf9a8104e5f92e236dde9</td>\n",
       "      <td>Emerging Infectious Diseases: A Historical and...</td>\n",
       "      <td>Gibril Ndow, J Radeino Ambe, Oyewale Tomori</td>\n",
       "      <td>Gibril Ndow, J Radeino Ambe, Oyewale Tomori</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\n\\nguide for their recognition and prevention...</td>\n",
       "      <td>History of yellow fever, G Augustine, , 1909; ...</td>\n",
       "      <td>[{'first': 'Gibril', 'middle': [], 'last': 'Nd...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'History...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2f57fc005f74c28402fe57d88144a855ee09efb2</td>\n",
       "      <td>Internet of Things (IoT) applications to fight...</td>\n",
       "      <td>Ravi Pratap Singh, Mohd Javaid, Abid Haleem, R...</td>\n",
       "      <td>Ravi Pratap Singh (Ambedkar National Institute...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Introduction\\n\\nThe Internet of Things (IoT) i...</td>\n",
       "      <td>Internet of things (IoT) applications in ortho...</td>\n",
       "      <td>[{'first': 'Ravi', 'middle': ['Pratap'], 'last...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Interne...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>238a7184d9d9876b23b92b957489b91c98365c39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\n\\npathogenesis of CIA, we have estabbshed T ...</td>\n",
       "      <td>We utilized the T-cell activating .superantige...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{'BIBREF1': {'ref_id': 'b1', 'title': \"We util...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  1d7234df1a19a0cf0b4038ceb31e7a97881cdbd7   \n",
       "1  0ced1f946cce007aa319a0ba38aef2c4b14dab0e   \n",
       "2  2a667552f5924857d6eaf9a8104e5f92e236dde9   \n",
       "3  2f57fc005f74c28402fe57d88144a855ee09efb2   \n",
       "4  238a7184d9d9876b23b92b957489b91c98365c39   \n",
       "\n",
       "                                               title  \\\n",
       "0  Virtual application of in situ simulation duri...   \n",
       "1  BMC Immunology Identification of a novel conse...   \n",
       "2  Emerging Infectious Diseases: A Historical and...   \n",
       "3  Internet of Things (IoT) applications to fight...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                             authors  \\\n",
       "0  Erich Hanel, E M , Monika Bilic, Kelly Hassall...   \n",
       "1  Yanbo Lv, Zhihua Ruan, Li Wang, Bing Ni, Yuzha...   \n",
       "2        Gibril Ndow, J Radeino Ambe, Oyewale Tomori   \n",
       "3  Ravi Pratap Singh, Mohd Javaid, Abid Haleem, R...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  Erich Hanel, E M , Monika Bilic, Kelly Hassall...   \n",
       "1  Yanbo Lv (Third Military Medical University, 3...   \n",
       "2        Gibril Ndow, J Radeino Ambe, Oyewale Tomori   \n",
       "3  Ravi Pratap Singh (Ambedkar National Institute...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Abstract\\n\\nThe coronavirus disease 2019 (COVI...   \n",
       "1  Abstract\\n\\nThe spike (S) protein is a major s...   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\nRÉSUMÉ L'arrivée de la pandémie causée par...   \n",
       "1  Background\\n\\nSevere acute respiratory syndrom...   \n",
       "2  \\n\\nguide for their recognition and prevention...   \n",
       "3  Introduction\\n\\nThe Internet of Things (IoT) i...   \n",
       "4  \\n\\npathogenesis of CIA, we have estabbshed T ...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  In situ simulation and its effects on patient ...   \n",
       "1  Identification of a novel coronavirus in patie...   \n",
       "2  History of yellow fever, G Augustine, , 1909; ...   \n",
       "3  Internet of things (IoT) applications in ortho...   \n",
       "4  We utilized the T-cell activating .superantige...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0  [{'first': 'Erich', 'middle': [], 'last': 'Han...   \n",
       "1  [{'first': 'Yanbo', 'middle': [], 'last': 'Lv'...   \n",
       "2  [{'first': 'Gibril', 'middle': [], 'last': 'Nd...   \n",
       "3  [{'first': 'Ravi', 'middle': ['Pratap'], 'last...   \n",
       "4                                                 []   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF0': {'ref_id': 'b0', 'title': 'In situ...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'Identif...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'History...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'Interne...  \n",
       "4  {'BIBREF1': {'ref_id': 'b1', 'title': \"We util...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_pdf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "text = list(clean_pdf_df['text'])[:5000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to Corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5000, 305977)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfidf = TfidfVectorizer()\n",
    "corpus_tfidf_matrix = tfidf.fit_transform(text)\n",
    "corpus_tfidf_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 305977)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "questions = [\n",
    "  'Development of a point-of-care test and rapid bed-side tests',\n",
    "  'Diagnosing SARS-COV-2 with Nucleic-acid based tech',\n",
    "  'Diagnosing SARS-COV-2 with antibodies'\n",
    "]\n",
    "test_tfidf_matrix = tfidf.transform(questions)\n",
    "test_tfidf_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_matrix = cosine_similarity(corpus_tfidf_matrix, test_tfidf_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5000, 3)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_articles_dic = {}\n",
    "sim_articles_list = []\n",
    "threshold = 0.25\n",
    "for i in range(sim_matrix.shape[0]):\n",
    "  for j in range(sim_matrix.shape[1]):\n",
    "    if sim_matrix[i][j] > threshold:\n",
    "      sim_articles_list.append([i, j, sim_matrix[i][j]])\n",
    "      sim_articles_dic[str(i)] = text[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "paragraph_list = []\n",
    "for key in sim_articles_dic:\n",
    "    text = sim_articles_dic[key]\n",
    "    text_dict=dict()\n",
    "    comp_list= text.split(\"\\n\\n\")\n",
    "    text_dict['index'] = key\n",
    "    for num in range(int((len(comp_list))/2)):\n",
    "        key_str=str(num)+'_'+str(comp_list[num*2])\n",
    "        key_str=key_str.strip()\n",
    "        text_dict[key_str]=str(comp_list[num*2+1])\n",
    "    paragraph_list.append(text_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[401, 2, 0.28245846198505487],\n",
       " [502, 2, 0.2817824050878491],\n",
       " [1301, 2, 0.2878332130702234],\n",
       " [1342, 2, 0.258659755115504],\n",
       " [1600, 2, 0.2577630044642292],\n",
       " [2456, 2, 0.25295082342974223],\n",
       " [3121, 2, 0.37539424715602265],\n",
       " [3207, 2, 0.33085976112834203],\n",
       " [3545, 2, 0.28369725740931473],\n",
       " [3981, 2, 0.2754481287807529],\n",
       " [4000, 2, 0.2834126407078308],\n",
       " [4098, 2, 0.26281561144474136],\n",
       " [4099, 2, 0.2603616069001906],\n",
       " [4476, 0, 0.2507892003265474]]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_articles_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def answer_question(question, answer_text):\n",
    "    '''\n",
    "    Takes a `question` string and an `answer_text` string (which contains the\n",
    "    answer), and identifies the words within the `answer_text` that are the\n",
    "    answer. Prints them out.\n",
    "    '''\n",
    "    # ======== Tokenize ========\n",
    "    # Apply the tokenizer to the input text, treating them as a text-pair.\n",
    "    input_ids = tokenizer.encode(question, answer_text,max_length=500\n",
    "                                )\n",
    "\n",
    "    # Report how long the input sequence is.\n",
    "    #print('Query has {:,} tokens.\\n'.format(len(input_ids)))\n",
    "\n",
    "    # ======== Set Segment IDs ========\n",
    "    # Search the input_ids for the first instance of the `[SEP]` token.\n",
    "    sep_index = input_ids.index(tokenizer.sep_token_id)\n",
    "\n",
    "    # The number of segment A tokens includes the [SEP] token istelf.\n",
    "    num_seg_a = sep_index + 1\n",
    "\n",
    "    # The remainder are segment B.\n",
    "    num_seg_b = len(input_ids) - num_seg_a\n",
    "\n",
    "    # Construct the list of 0s and 1s.\n",
    "    segment_ids = [0]*num_seg_a + [1]*num_seg_b\n",
    "\n",
    "    # There should be a segment_id for every input token.\n",
    "    assert len(segment_ids) == len(input_ids)\n",
    "\n",
    "    # ======== Evaluate ========\n",
    "    # Run our example question through the model.\n",
    "    start_scores, end_scores = model(torch.tensor([input_ids]), # The tokens representing our input text.\n",
    "                                    token_type_ids=torch.tensor([segment_ids])) # The segment IDs to differentiate question from answer_text\n",
    "\n",
    "    # ======== Reconstruct Answer ========\n",
    "    # Find the tokens with the highest `start` and `end` scores.\n",
    "    answer_start = torch.argmax(start_scores)\n",
    "    answer_end = torch.argmax(end_scores)\n",
    "    \n",
    "    \n",
    "    # Get the string versions of the input tokens.\n",
    "    tokens = tokenizer.convert_ids_to_tokens(input_ids)\n",
    "\n",
    "    # Start with the first token.\n",
    "    answer = tokens[answer_start]\n",
    "\n",
    "    # Select the remaining answer tokens and join them with whitespace.\n",
    "    for i in range(answer_start + 1, answer_end + 1):\n",
    "        \n",
    "        # If it's a subword token, then recombine it with the previous token.\n",
    "        if tokens[i][0:2] == '##':\n",
    "            answer += tokens[i][2:]\n",
    "        \n",
    "        # Otherwise, add a space then the token.\n",
    "        else:\n",
    "            answer += ' ' + tokens[i]\n",
    "            \n",
    "    s_scores = start_scores.detach().numpy().flatten()\n",
    "    e_scores = end_scores.detach().numpy().flatten()\n",
    "    #print('score:'+(start_scores)+\"; \"+str(end_scores))\n",
    "    #print('score:'+str(max(s_scores))+\"; \"+str(min(e_scores)))\n",
    "    #print(str(tensor[torch.argmax(start_scores)]))\n",
    "    #print('Answer: \"' + answer + '\"')\n",
    "    #[answer,str(max(s_scores)),len(input_ids)]\n",
    "    return answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Results\\n\\nExpression of SARS-CoV-2 S protein. To enhance expression of the S protein of SARS-CoV-2 in mammalian cells, a codonoptimized cDNA encoding the S protein and 3xFLAG tag was synthesized, and to facilitate incorporation of S protein into lentiviral pseudovirons, the last 19 amino acids containing an endoplasmic reticulum (ER)-retention signal from the cytoplasmic tail of the S protein was removed (Fig. 1a) . The construct was named SARS-CoV-2 S. HEK293T cells were transfected with SARS-CoV-2 S plasmid and expression of SARS-CoV-2 S protein was determined by western blot. There were two major bands, 180 kDa, and 90 kDa, detected by mouse anti-FLAG M2 antibody (Fig. 1b, lane 2) , reflecting the full-length and cleaved S proteins, respectively. The band above 250 kDa likely results from dimeric or trimeric S proteins. Consistent with our previous report 29 , MERS-CoV S protein was detected by polyclonal goat anti-MHV S antibodies AO4 (Fig. 1c) . AO4 also detected SARS-CoV-2 and SARS-CoV S proteins, suggesting the presence of a conserved immunogenic epitope among all four different CoVs. This presumably linear epitope is likely in S2 29 . S1 subunits of SARS-CoV-2 and SARS-CoV share almost 64% in amino acid identities. Nevertheless, SARS-CoV-2 S protein was barely detected by rabbit polyclonal anti-SARS S1 antibodies T62 (Fig. 1d) , suggesting that the major epitope(s) for T62 antibodies include nonconserved regions of S1. The SARS-CoV-2 S protein was not detected by either a monoclonal anti-SARS S1 antibody ( Fig. 1e ) or anti-MERS S2 antibody (Fig. 1f) .Pseudovirion incorporation of SARS-CoV-2 S protein. The efficiency of SARS-CoV-2 S protein incorporation into pseudovirions was evaluated using monoclonal mouse anti-FLAG M2 antibody and polyclonal goat anti-MHV S antibody AO4. While majority of SARS-CoV S proteins incorporated into pseudovirons were full-length, at 180 kDa (Fig. 1h, i) , most of SARS-CoV-2 S proteins on lentiviral pseudovirions were cleaved, about 90 kDa (Fig. 1g, h) , likely reflecting presence of extra furin site (R682-R683-A684-R685, Fig. 1a ) between S1 and S2 in SARS-CoV-2 S protein. Consistent with the results in cell lysate (Fig. 1d) , SARS-CoV S protein, but not SARS-CoV-2 S protein, in pseudovirions was readily detected by using polyclonal rabbit anti-SARS S1 antibodies T62 (Fig. 1i) .Human ACE2 is a receptor for SARS-CoV-2. Next, we determined whether SARS-CoV-2 S pseudovirions were able to transduce human, monkey, and bat cells. VSV-G pseudovirons were used as a positive control, whereas bald particles with no spike proteins (mock) served as a negative control. As expected, all cell types were effectively transduced by VSV-G pseudovirons (Fig. 2a) . Compared to mock control, SARS-CoV-2 S pseudovirions showed an over 500-fold increase in luciferase activities in Calu3 cells, at a level similar to SARS-CoV S pseudovirions (Fig. 2a) . Huh7 and Vero 81 cells also gave about 10-fold increase in luciferase activities when transduced by SARS-CoV-2. Transduction of LLCMK2 cells was higher with SARS-CoV S pseudovirions than with SARS-CoV-2 S pseudovirions (Fig. 2a) , suggesting that there might be some differences on virus entry on LLCMK2 cells mediated by S proteins between SARS-CoV-2 and SARS-CoV. We then investigated whether any known CoV receptors might be used by SARS-CoV-2 S protein as entry receptor. The SARS-CoV-2 S pseudovirons were used to transduce BHK cells stably expressing human aminopeptidase N (hAPN), the receptor for human CoV 229E, 293 cells stably expressing hACE2 (293/hACE2), the receptor for SARS-CoV, and HeLa cells stably expressing hDPP4 (HeLa/hDPP4), the receptor for MERS-CoV. While BHK/hAPN and HeLa/hDPP4 cells were not susceptible for the transduction of SARS-CoV-2 S pseudovirions, 293/hACE2 cells were highly transduced by SARS-CoV-2 S pseudovirions, consistent with hACE2 as the receptor for SARS-CoV-2 13 . We then determined whether SARS-CoV-2 S protein could directly bind to hACE2 protein. HEK 293T cells transiently expressing SARS-CoV-2 S protein were incubated with soluble hACE2 and analyzed by flow cytometry. As shown in Fig. 2c , SARS-CoV-2 S protein bound to soluble hACE2 at a level similar to SARS-CoV S protein, although the mean fluorescence intensity (MFI) for SARS-CoV-2 S protein was slightly lower than SARS-CoV S protein. To further investigate if hACE2 is the receptor for SARS-CoV-2, we performed inhibition experiments using soluble hACE2. Soluble hACE2 proteins were pre-incubated with SARS-CoV-2 S pseudovirons on ice for 1 h, then virus-protein mixture was added onto 293/hACE2 cells. Entry of SARS-CoV-2 S pseudovirions was significantly prevented by pre-incubation of soluble hACE2 at both 10 μg/ml and 50 μg/ml (Fig. 2d) , further supporting the notion that hACE2 is the receptor and soluble hACE2 might be used as a therapeutic inhibitor against SARS-CoV-2 infection.The SARS-CoV-2 enters 293/hACE2 cells through endocytosis. The majority of S proteins on SARS-CoV-2 S pseudovirions are cleaved (Fig. 1g, h) . We next determined whether SARS-CoV-2 S pseudovirons entered cells through endocytosis or cell surface. HEK 293/hACE2 cells were treated with lysosomotropic agents, ammonia chloride and bafilomycin A, and their effect on virus entry was evaluated. Consistent with previous reports, 20 mM NH 4 Cl and 100 nM bafilomycin A decreased entry of SARS-CoV S and VSV-G pseudovirions by over 99%, compared to no treatment control. More than 98% reduction in transduction on 293/hACE2 cells by SARS-CoV-2 S pseudovirions was also shown when the cells were incubated with either NH 4 Cl or bafilomycin A (Fig. 3a) , indicating that SARS-CoV-2 S pseudovirions enter 293/hACE2 cells mainly through endocytosis, despite that its spike proteins were cleaved.PIKfyve and TPC2 critical for SARS-CoV-2 entry. Phosphoinositides play many essential roles in endocytosis. Among them, one is phosphatidylinositol-3,5-bisphosphate (PI(3,5)P 2 ), which regulates early endosome to late endosome dynamics 33, 34 . Phosphatidylinositol 3-phosphate 5-kinase (PIKfyve) is the main enzyme synthesizing PI(3,5)P 2 in early endosome. HEK 293/ hACE2 cells were treated with apilimod, a potent inhibitor for PIKfyve 35 Monoclonal anti-SARS S1Polyclonal anti-SARS S1Polyclonal anti-SARS S1 polyclonal anti-MHV S AO4 Anti-Flag M2 NATURE COMMUNICATIONS | https://doi.org/10.1038/s41467-020-15562-9 ARTICLE NATURE COMMUNICATIONS | (2020) 11:1620 | https://doi.org/10.1038/s41467-020-15562-9 | www.nature.com/naturecommunications dependent manner (Fig. 3b) , whereas it had no effect on entry of VSV-G pseudovirions, which occurs in early endosomes. Similar inhibitory effects were observed when HeLa/hDPP4 cells and HeLa cells stably expressing mouse carcinoembryonic antigen related cell adhesion molecule 1a (mCEACAM1a) (HeLa/mCEACAM1a) were treated with apilimod and transduced with MERS-CoV and MHV S pseudovirions (Fig. 3b) , respectively. Moreover, infection of live MHV on 17Cl.1 cells was also strongly inhibited by apilimod treatment (Fig. 3c ). No significant cell toxicity was observed on apilimod at any concentration tested (Fig. 3c) . We then determined whether apilimod could inhibit entry of SARS-CoV-2 S pseudovirions on 293/hACE2. As expected, apilimod treatment inhibited entry of SARS-CoV-2 S pseudovirions in a dose dependent manner (Fig. 3d) . Similar effects were shown when 293/ hACE2 cells were treated with YM201636, another PIKfyve inhibitor (Fig. 3e) . These results suggested that PIKfyve might be a potential general drug target for viruses that enter cells through endocytosis. Two pore channel subtype 2 (TPC2) and TRPML1 in lysosome are two major downstream effectors of PI(3,5)P 2 36 . While blocking TPC2 activity by tetrandrine, an inhibitor for TPC2 37 , decreased entry of SARS-CoV-2 S pseudovirions ( Fig. 3f) , treatment of cells with 130, a TRPML1 inhibitor, had no effect (Supplementary Fig. 1 ), indicating that TPC2, not TRPML1, is important for SARS-CoV-2 entry.Effect of cathepsin inhibitors on SARS-CoV-2 entry. Protease activation on S protein is an important step for coronavirus entry, and cathepsins in lysosome have been shown to be critical for SARS-CoV and MERS-CoV entry through endocytosis. To investigate whether cathepsins are required for SARS-CoV-2 entry, HEK 293/hACE2 cells were treated with either a broad inhibitor for cathepsin B, H, L, and calpain (E64D), cathepsin L inhibitor (SID 26681509), or cathepsin B inhibitor (CA-074). VSV-G pseudovirions were used as a negative control, since virus entry mediated by VSV-G does not require protease activation. E64D treatment of 293/hACE2 cells reduced entry of SARS-CoV-2 S pseudovirions by 92.5%, indicating that at least one of cathepsins or calpain might be required for SARS-CoV-2 entry (Fig. 4a) . While cathepsin B inhibitor treatment did not show any marked effect on virus entry, cathepsin L inhibition treatment decreased entry of SARS-CoV-2 S pseudovirions into 293/hACE2 by over 76% (Fig. 4a) , suggesting that cathepsin L should be essential for priming of SARS-CoV-2 S protein in lysosome for entry into 293/hACE2 cells.Cell-cell fusion mediated by SARS-CoV-2 S protein. Type II membrane serine proteases (TMPRSS)-mediated cleavage can activate the fusion potential of SARS-CoV and MERS-CoV S protein and induce receptor-dependent syncytium formation 28, 29, 38, 39 .To test whether SARS-CoV-2 S protein cleaved by these cellular proteases could trigger syncytium formation, we overlaid HEK 293T cells expressing SARS-CoV-2 S or control proteins on 293/ hACE2 cells in presence or absence of proteases. Expression of TMPRSS 2, 4, 11 A, 11D, and 11E on 293/hACE2 cells enhanced SARS-CoV-2 S protein-mediated cell-cell fusion similarly to SASR-CoV S ( Supplementary Fig. 2 ). Next we evaluated the role of trypsin. Consistent to our previous report 23 , addition of trypsin triggered syncytium formation on 293/hACE2 cells by SARS-CoV S protein after 4 h incubation (Fig. 4b, c) . Large syncytia were formed at a level similar to SARS-CoV S proteins when SARS-CoV-2 S protein expressing 293T cells were added onto 293/hACE2 cells with trypsin. Of note, SARS-CoV-2, but not SARS-CoV, S proteins induced syncytia formation on 293/hACE2 cells even in the absence of trypsin (Fig. 4c) , suggesting that SARS-CoV-2 S proteins could be triggered upon the receptor binding without exogenous protease priming or activation. During 2013-2016 Ebola virus outbreak, A82V and T544I mutations in glycoprotein (gp) of Ebola virus were linked to increase of virus transmissibility [40] [41] [42] . Further study showed that A82V and T544I mutation decreases the thermostability of gp while they increases virus infectivity in cell culture 42 . We then investigated the thermostability of SARS-CoV-2 S protein.Compared to SARS-CoV S protein, SARS-CoV-2 S protein was less stable, requiring significant shorter time and lower temperature to be inactivated (Fig. 4d, e) . The native S protein is metastable, and there is an energy barrier that prevents it from undergoing conformational change before triggering, SARS-CoV-2 S protein might decrease its energy barrier by reducing its thermostability. This might contribute to high-transmission efficiency of SARS-CoV-2.Effect of anti-SARS S1 antibodies T62 on SARS-CoV-2 S protein.Since SARS-CoV-2 S reacted weakly with polyclonal rabbit anti-SARS S1 antibodies T62 in western blot (Fig. 1d) , we investigated whether SARS-CoV-2 S protein in native conformation could be recognized by anti-SARS S1 antibodies T62.HEK293T cells transiently expressing SARS-CoV-2 S proteins were incubated with polyclonal anti-SARS S1 antibodies T62 and analyzed by flow cytometry. SARS-CoV S was used as a positive control. As expected, expression of SARS-CoV S proteins on 293T cell surface were readily detected by polyclonal anti-SARS S1 antibodies T62 (Fig. 5a ). In contrast, only low level of binding of SARS-CoV-2 S protein to polyclonal rabbit anti-SARS S1 antibodies T62 was detected (Fig. 5a) . To further delineate where the conformational epitopes for this polyclonal antibodies are located, two new constructs were generated, a SARS-CoV S protein backbone with RBD from SARS-CoV-2 (SARS-CoV S/nRBD) and a SARS-CoV-2 S protein backbone with RBD from SARS-CoV (SARS-CoV-2 S/sRBD). Replacement of SARS-CoV RBD with SARS-CoV-2 RBD (SARS-CoV S/nRBD) decreased binding of polyclonal antibodies T62 to S protein, whereas substitution of SARS-CoV-2 RBD with SARS-CoV RBD (SARS-CoV-2 S/sRBD) increased the affinity of S protein to polyclonal antibodies T62 (Fig. 5a) . All chimera proteins were expressed at levels similar to SARS-CoV-2 S protein (Fig. 5b, top panel) . These results suggest that at least one major conformational epitope for polyclonal antibodies T62 might be located in RBD where the sequences differ between SARS-CoV and SARS-CoV-2 (Fig. 5c ). Of note, while SARS-CoV-2 S/sRBD bound to polyclonal antibodies T62 better than SARS-CoV S/nRBD in flow cytometry, the signal for SARS-CoV S/nRBD was stronger than SARS-CoV-2 S/sRBD in western blot by polyclonal antibodies T62, indicating that major linear epitopes of polyclonal antibodies T62 were located on SARS-CoV NTD. Among residues in SARS-CoV RBD critical for receptor binding and virus entry 43, 44 , seven of them are different between SARS-CoV and SARS-CoV-2 RBDs (Fig. 5c) . We then mutated each of them in SARS-CoV S protein to the corresponding residues in SARS-CoV-2 S. With exception of SARS-CoV mutant Y442L, which had reduced protein expression ( Supplementary Fig. 3) , binding of SARS-CoV S mutants to polyclonal rabbit anti-SARS S1 antibodies T62 was not affected (Fig. 5d ). This suggests that these residues might not be direct binding sites for polyclonal antibodies T62. Next, we tested whether polyclonal rabbit anti-SARS S1 antibodies T62 could inhibit entry of SARS-CoV-2 S pseudovirions. CoV S protein pseudovirions were incubated with polyclonal antibodies T62 on ice for 1 h, and their transduction was measured according to luciferase activities. As expected, polyclonal antibodies T62 neutralized SARS-CoV S pseudovirion effectively in a dose dependent manner (Fig. 5e ). In contrast, even at a concentration of 50 μg/ml, polyclonal antibodies T62 did not have marked effect on transduction by SARS-CoV-2 S proteins (Fig. 5e) .Cross-neutralization of SARS and COVID-19 patient sera. As rabbit polyclonal anti-SARS S1 antibodies did not show significant neutralization activity against SARS-CoV-2 S pseudovirions, we then asked whether sera from recovered SARS and COVID-19 patients could neutralize SARS-CoV S and SARS-CoV-2 S pseudovirions. We obtained sera from one recovered SARS patient and five recovered COVID-19 patients and tested their neutralization activities against transduction on 293/hACE2 cells by SARS-CoV and SARS-CoV-2 S pseudovirions. Serum from recovered SARS patient showed strong inhibition on transduction by SARS-CoV S pseudovirions, and it had modest neutralization activity against SARS-CoV-2 S pseudovirons ( Fig. 6a and Table 1 ). In contrast, sera from all five COVID-19 patients neutralized SARS-CoV-2 S pseudovirions effectively, and none of them had an effect on transduction by SARS-CoV S pseudovirions ( Fig. 6b and Table 1 ). Discussion About 70% of the emerging pathogens infecting humans originate from animals, and CoVs are among the forefronts of these pathogens 45 . The newly emerged SARS-CoV-2 infects human and causes severe pneumonia, and as of 10 February, 2020, the current outbreak has spread to 25 countries with over 40,000 confirmed cases and 900 deaths 46 . However, little is known about its biology. Since the virus is categorized as a biosafety level 3 (BSL3) agent, according to WHO guideline, we developed a pseudotype system with S protein of SARS-CoV-2 to study virus entry in BSL2 settings. Understanding how SARS-CoV-2 enters cell will provide valuable information for virus pathogenesis, vaccine design, and drug target. Polyclonal anti-SARS S1 antibodies T62 Utilizing this pseudotype system, we screened a panel of human and monkey cell lines for their susceptibility of SARS-CoV-2 S-mediated transduction. In line with SASR-CoV-2 causing respiratory infections in humans, we found that human lung cancer cell line Calu3 is highly susceptible to SARS-CoV-2 Smediated entry. LLCMK2 cells, a rhesus monkey kidney epithelium cell line, exhibited different susceptibility to SARS-CoV S and SARS-CoV-2 S transduction, but the reasons for this are currently not known and require further investigation. Both S proteins use hACE2 as the receptor for binding and entry 13, 21 , which we further confirmed by flow cytometry analysis and competitive inhibition experiment using soluble hACE2 in this study (Fig. 3b, c) . While full-length S proteins of SARS-CoV-2 and SARS-CoV S share almost 76% identities in amino acid sequences, the NTDs show only 53.5% of homology. Of note, NTDs of different CoVs S proteins have been shown to bind different sugar. While NTD of MERS-CoV prefers α2,3-linked sialic acid over α2,6-linked sialic acid 47 , NTDs of human CoVs OC43 and HKU1 bind to 9-O-acetylated sialic acids 48, 49 . No sugar binding has been reported for NTD of SARS-CoV. Whether or not NTD of SARS-CoV-2 binds to sugar and whether sugar binding of NTD affects virus entry remains to be determined.Successful conformational changes of S proteins, leading to membrane fusion, not only require receptor binding, but also need appropriate protease activation. There is a furin site between S1 and S2 (amino acids 682-685, RRAR) subunits in SARS-CoV-2 S protein, similar to what happens in high-pathogenic influenza viruses. Whether the presence of this furin site has any effect on viral pathogenesis and virus spreading among humans remains to be determined. Although majority of SARS-CoV-2 S proteins in pseuodvirions were cleaved, endocytosis is the main entry pathway on 293/hACE2 cell. In our experiments, lysosomal cathepsin L, not B, appeared critical for SARS-CoV-2 S protein activation, similar to what was reported for SARS-CoV and MERS-CoV 28, 29, 50 . Trypsin could also activate SARS-CoV-2 S protein efficiently and induce large syncytium formation. To our surprise, we found that, even without trypsin, SARS-CoV-2 S protein could trigger syncytium formation on 293/hACE2 cells, a phenomena similar to what was Fig. 5 Characterization of polyclonal rabbit anti-SARS S1 antibodies T62. a Binding of polyclonal rabbit anti-SARS S1 antibodies T62 to SARS-CoV-2, SARS-CoV S, and chimeric S proteins. HEK293T cells transiently expressing either SARS-CoV-2 S, SARS-CoV S, SARS-CoV S/nRBD, or SARS-CoV-2 S/ sRBD proteins were incubated with polyclonal rabbit anti-SARS-CoV S1 antibody T62 for 1 h on ice, followed by a FITC-conjugated secondary antibody, then cells were analyzed by flow cytometry. Experiments were done three times and one representative is shown. b Expression of SARS-CoV-2 S, SARS-CoV S, or chimeric S proteins on 293T cells. Cells from panel A were lyzed and blotted with anti-FLAG M2 antibody and polyclonal anti-SARS S1 antibody T62. c Amino acid sequence alignment of SARS-CoV and SARS-CoV-2 S RBDs. Stars (*) indicate the seven critical residues different between SARS-CoV-2 and SARS-CoV RBDs. d Binding of polyclonal rabbit anti-SARS S1 antibodies T62 to mutant SARS-CoV S proteins. e Neutralization of SARS-CoV-2 S and SARS-CoV S pseudovirions by polyclonal rabbit anti-SARS S1 antibody T62. Pseudovirons were pre-incubated with serially diluted polyclonal rabbit anti-SARS S1 antibodies T62 on ice, then virus-antibody mixture was added on 293/hACE2 cells. Pseudoviral transduction was measured 40 h later. Experiments were done in triplicates and repeated twice, and one representative is shown. Error bars indicate SEM of technical triplicates. Source data are provided as a Source Data file. SARS-CoV and SARS-CoV-2 S pseudovirions were pre-incubated with serially diluted patient sera for 1 h on ice, then virus-antibody mixture was added on 293/hACE2 cells. Cells were lysed 40 h later and pseudovirus transduction was measured. The highest dilution of the serum sample that decreased transduction by 50% or more was considered to be positive. Source data are provided as a Source Data file.\\n\\n% of CTRL transduction\\n\\nobserved in MHV S protein in certain aspect 51 . This led us to speculate that SARS-CoV-2 S protein is capable of triggering protease-independent and receptor-dependent syncytium formation. Such a mechanism might enhance virus spreading through cell-cell fusion and this might partially explain rapid progress of the diseases 10, 11 . However, we have not excluded a potential role for other endogenous proteases in HEK293T cells for syncytium formation and we have also not evaluated entry into other cell types. This requires further research. Recent studies showed that early to late endosome maturation is regulated by PI(3,5)P 2 , and inhibition of PIKfyve, the key enzyme synthesizing PI(3,5)P 2 , and TPC2, a downstream effector in lysosome, significantly reduced virus entry of MERS-CoV 52 . We confirmed this in this study, and further showed that blocking PIKfyve and TPC2 also strongly inhibited entry mediated by SARS-CoV-2 S protein, indicating that PI(3,5)P 2 pathway might be considered as potential general drug target for CoV infection.CoV S protein is one of the key components determining virus virulence, tissue tropism and host range, and it is also a main target for neutralizing antibodies and vaccine design. Although the S proteins of SARS-CoV-2 and SARS-CoV are highly homologous, polyclonal rabbit anti-SARS S1 antibodies T62 did not bind to SARS-CoV-2 S protein well, and poorly neutralized SARS-CoV-2 S protein-mediated virus entry. Further analysis reveals that major immune-epitopes for T62 antibodies likely lie in the region of RBD (Fig. 4a) . We further evaluated crossneutralization of SARS-CoV and SARS-CoV-2 using convalescent sera from SARS and COVID-19 patients. We only saw moderate cross-neutralization activities between SARS-CoV-2 and SARS-CoV convalescent sera, suggesting that those previously recovered from SARS-CoV infection may not be fully protected against SARS-CoV-2 infection, and vice versa.While this manuscript was under review, Hoffmann et al. 53 reported that human ACE2 is the entry receptor of SARS-CoV-2, and that serine protease TMPRSS2 is important for SARS-CoV-2 S activation, consistent with our findings. In agreement with our work, sera from convalescent SARS patients neutralized SARS-CoV-2 S pseudovirion entry with lower efficiency than SARS-CoV S pseudovirion entry, showing moderate cross-neutralization.In conclusion, we demonstrated that SARS-CoV-2 S protein entry on 293/hACE2 cells is mainly mediated through endocytosis, and that PIKfyve, TPC2, and cathepsin L are critical for virus entry. We further found that SARS-CoV-2 S protein could trigger syncytia in 293/hACE2 cells independent of exogenous protease. Finally, there was limited cross-neutralization activity between convalescent sera from SARS and COVID-19 patients. Our findings provide potential targets for development of drugs and vaccines against this newly emerging lineage B beta-CoV.\\n\\nMethods\\n\\nConstructs and plasmids. Codon-optimized cDNA (sequence shown in Supplementary Table 2 ) encoding SARS-CoV-2 S glycoprotein (QHU36824.1) with Cterminal 19 amino acids deletion was synthesized and cloned into eukaryotic cell expression vector pCMV14-3×Flag between the Hind III and BamH I sites. Plasmids encoding SARS-CoV S glycoprotein, MERS-CoV S glycoprotein and MHV S glycoprotein (sequences shown in Supplementary Table 2) were constructed by inserting DNA fragment encoding codon-optimized SARS-CoV S protein (AAP13441.1) lacking the last 19 amino acids (aa), MERS-CoV S protein (AFS88936.1) lacking the last 16 aa but with a FLAG tag at the C terminus, or fulllength MHV S protein (AAU06356.1) into pcDNA3.1between BamH I and Not I sites, respectively. The VSV-G encoding plasmid and lentiviral packaging plasmid psPAX2 were obtained from Addgene (Cambridge, MA). The pLenti-GFP lentiviral reporter plasmid that expresses GFP and luciferase was generously gifted by Fang Li, Duke University. All primers used in this study are listed in Supplementary Table 1 . Production of SARS-CoV-2 S pseudovirions and virus entry. Pseudovirions were produced by co-transfection 293T cells with psPAX2, pLenti-GFP, and plasmids encoding either SARS-CoV-2 S, SARS-CoV S, VSV-G, or empty vector by using polyetherimide (PEI). The supernatants were harvested at 40, 64 h post transfection, passed through 0.45 μm filter, and centrifuged at 800 × g for 5 min to remove cell debris. To transduce cells with pseudovirions, cells were seeded into 24-well plates and inoculated with 500 μl media containing pseudovirions. After overnight incubation, cells were fed with fresh media. About 40 h post inoculation, cells were lysed with 120 μl medium containing 50% Steady-glo (promega) at room temperature for 5 min. The transduction efficiency was measured by quantification of the luciferase activity using a Modulus II microplate reader (Turner Biosystems, Sunnyvale, CA, USA). All experiments were done in triplicates, and repeated at least twice or more.Detection of S protein of SARS-CoV-2 by western blot. The spike glycoprotein of SARS-CoV-2 in cells and on pseudovirions were detected by using western blot. Briefly, cells transfected with plasmids encoding spike glycoprotein of SARS-CoV-2, SARS-CoV, MERS-CoV, and MHV were lysed at 40 h post transfection by RIPA buffer (20 mM Tris-HCl pH 7.5, 150 mM NaCl, 1 mM EDTA, 0.1% SDS, 1% NP40, 1×protease inhibitor cocktail). To pellet down pseudovirions, the viral supernatants were centrifuged at 25,000 rpm for 2 h in a Beckman SW41 rotor at 4°C through a 20% sucrose cushion, and virus pellets were resuspended into 30 μl RIPA buffer. The samples were boiled for 10 min and separated in a 10% SDS-PAGE gel and transferred to nitrocellulose filter membranes. After blocked by 5% milk, the membranes were blotted with primary antibodies, followed by incubated with horseradish peroxidase (HRP) conjugated secondary antibodies (1:5000) and visualized with Chemiluminescent Reagent (Bio-Rad). MHV S proteins were detected using polyclonal goat anti-MHV S antibody AO4 (1:2000); SARS-CoV S proteins were blotted with either polyclonal anti-SARS S1 antibodies T62 (1:2000) (Sinobiological Inc, Beijing, China) or mouse monoclonal against SARS S1 antibody MM02 (1:1000) (Sinobiological Inc, Beijing, China), MERS-CoV and SARS-CoV-2 S proteins were detected using mouse monoclonal anti-MERS S (1:1000) (Sinobiological Inc, Beijing, China) and anti-FLAG M2 antibody (1:1000) (Sigma, St. Louis, MO, USA), respectively.Effects of agents and inhibitors on pseudovirion entry. HEK 293/hACE2 cells were pretreated with either lysosomotropic agents (endosome acidification inhibitor NH 4 Cl and bafilomycin A; PIKfyve inhibitor apilimod and YM201636; calcium channel blocker tetrandrine) or cathepsin inhibitors (cathepsin L inhibitor SID26881509, cathepsin B inhibitor CA-074, and cathepsin inhibitor E64D) for 1 h at 37°C, then spin-inoculated with pseudovirions in the presence of inhibitor at 1200 × g for 30 min. After overnight incubation, cells were fed with fresh medium without inhibitor. Cells were lysed at 48 h post inoculation and their luciferase activity was measured.Flow cytometric analysis of S protein expression. Briefly, 293T cells were transfected with 2 μg of plasmids encoding either SARS-CoV-2 S, SARS-CoV S or MERS-CoV S protein using PEI. Forty hours later, cells were detached by using PBS with 1 mM EDTA. After washing, cells were incubated with polyclonal rabbit anti-SARS S1 antibodies T62 (1:200 dilution) (Sinobiological Inc., Beijing, China) for 1 h on ice, followed by Alexa Fluor 488-conjugated goat anti-rabbit IgG (1:200) (ZSGB-Bio LLC, Beijing, China) for 1 h. Cells were then analyzed by flow cytometry.MTT assay. MTT assay was used to assess cell viability. Briefly, cells were seeded into a 96-well plate at a cell density of 3000 per well and allowed to adhere for 24 h, followed by treatment with serially diluted inhibitors for 12 h. Two microliters of 5 mg/ml MTT was added to the medium 24 h later and incubated for 4 h at 37°C. After removing the culture medium, 100 µl of DMSO was added. The absorbance was measured at 570 nm using a microplate spectrophotometer (Multiskan FC, Thermo Scientific). All experiments were performed in triplicate, and the cell viability of cells was calculated as the ratio of each experimental condition to the control.Cell-cell fusion assay. HEK-293T cells were co-transfected with plasmids encoding CoV S glycoprotein and eGFP. For trypsin-dependent cell-cell fusion, cells were detached with trypsin (0.25%) at 40 h post transfection and overlaid on an 80% confluent monolayer of 293/hACE2 cells at a ratio of approximately one Sexpressing cell to three receptor-expressing cells. For trypsin-independent cell-cell fusion, cells were detached with 1 mM EDTA and overlaid on 293/hACE2 cells. After 4 h incubation, images of syncytia were captured with a Nikon TE2000 epifluorescence microscope running MetaMorph software (Molecular Devices). All experiments were done in triplicate and repeated at least three times.Expression and purification of soluble human ACE2. DNA fragments encoding residues 19-615 of human hACE2 with N-terminal FLAG and 6×his tags were cloned between Sal I and Hind III of modified pFASTBac1 vector with gp67 signal peptide. The soluble receptors were expressed in High Five insect cells using the bac-to-bac system (Invitrogen) and purified using nickel affinity and ion-exchange chromatography.Soluble hACE2-binding assay. HEK-293T cells were transfected with plasmids encoding SARS-CoV-2 S, SARS-CoV S, or MERS-CoV S protein. Cells were detached with 1 mM EDTA 40 h post transfection, washed twice with 3% FBS in 1×PBS, and incubated with 5 μg/ml soluble hACE2 for 1 h on ice. After washing three times with 3% FBS in 1×PBS, cells were incubated with polyclonal goat antihuman ACE2 antibody (1:200) (R&D Systems, MN, USA) for 1 h, followed by FITC-conjugated rabbit anti-goat secondary antibody (1:500) (Jackson Immu-noResearch, PA, USA). After washing, cells were then analyzed by flow cytometry.Soluble hACE2 inhibition assay. Briefly, lentiviruses pseudotyped with SARS-CoV-2 S, SARS-CoV S or VSV-G were pre-incubated with serially diluted soluble hACE2 for 1 h on ice. The mixture were then added on 293/hACE2 cells, followed by centrifugation inoculation for 1 h at room temperature. Cells were fed with fresh medium 6 h later and lysed at 48 h post inoculation. Pseudoviral transduction was measured according to luciferase activities.Pseudovirus neutralization assay. SARS-CoV S, SARS-CoV-2 S, and VSV-G pseudovirions were pre-incubated with serially diluted either polyclonal rabbit anti-SARS S1 antibodies T62 or patient sera for 1 h on ice, then virus-antibody mixture was added onto 293/hACE2 cells in a 96-well plate. After 6 h incubation, the inoculum was replaced with fresh medium. Cells were lysed 40 h later and pseudovirus transduction was measured as previously described. Prior to experiments, patient sera were incubated at 56°C for 30 min to inactivate complement.Serum collection. Five patients were hospitalized with pneumonia in Hengshui Third People's Hospital, Hengshui, Heibei province, China, and their respiratory specimens were collected for coronavirus detection and were positive for SARS-CoV-2. After all patients recovered, their sera were collected right before discharge with informed consent. Serum from a recovered SARS patient was also collected at two years after recovery with informed consent.Reporting summary. Further information on research design is available in the Nature Research Reporting Summary linked to this article.\\n\\nData availability\\n\\nThe data supporting the findings of this study are available from the authors upon request. The source data underlying Figs. 1b-j, 2a, b, d, 3a-f, 4a , c-e, 5b, e and 6a, b and Table 1 and Supplementary Figs. 1 and 3 are provided as a Source Data file.Received: 13 February 2020; Accepted: 16 March 2020;\\n\\n\""
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text[3121]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "human ace2\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diffractable crystals of the sars - cov - 2 - ctd / hace2 complex\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "hace2\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "gisaid : epi _ isl _ 402119\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "spr\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "401\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "specific antibody response against the virus in both patients and experimentally infected primates\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "vaccines\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "vaccines for sars cov . 27\n",
      "502\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "develop rapid and accurate diagnostic tests for early diagnosis of sars cov infection\n",
      "1301\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "structural analysis\n",
      "1301\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1301\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1342\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP] there are currently no data on the specific role of either humoral or cellular immunity or innate immunity in patients recovering from covid - 19 . only highly specialized laboratories are able to conduct experiments to investigate immune responses against hla class - i and class - iirestricted viral epitopes mediated by cd8 + and cd4 + t lymphocytes , respectively , to confirm the conjecture of a cellular ( rather than humoral ) immunity against sars - cov - 2 . moreover , the t lymphocytes responsible for clinically relevant antiviral immune responses have high chances to be locally present in , or close to , respiratory epithelia but have comparatively low chances to be detectable in peripheral blood [ 57 ] [ 58 ] [ 59 ] . it is well possible that the exclusive detection of humoral immunity ( antibodies against sars - cov - 2 ) leads to an underestimation of the anti - sars - cov - 2 immune responses\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "1600\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "2456\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3121\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3121\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3121\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3121\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP] coronaviruses ( covs ) belong to the subfamily orthocoronavirinae in the family coronaviridae , order nidovirales . there are four genera within the subfamily orthocoronavirinae , namely alphacoronavirus ( α - cov ) , betacoronavirus ( β - cov ) , gammacoronavirus ( γ - cov ) and deltacoronavirus ( δ - cov ) [ 1 , 2 ] . the cov genome is an enveloped , positive - sense , single - stranded rna with a size varying between 26 kb and 32 kb , the largest genome of known rna viruses . both αand β - cov genera are known to infect mammals , whilst δand γ - covs infect birds . two recent outbreaks of viral pneumonia caused by β - covs are severe acute respiratory syndrome ( sars ) and middle east respiratory syndrome ( mers ) . , an outbreak of sars was first reported in china and then spread quickly worldwide , resulting in hundreds of deaths with an 11 % mortality rate [ 3 , 4 ] . in marise recent progress in sars - cov - 2 to provide a framework for the prevention and treatment of covid - 19\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "the disease was been confirmed to be an acute respiratory infection caused by a novel cov .\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "prevention\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "liu\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "asymptomatic or mild cases can also spread sars - cov - 2 between humans .\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "ct\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "suspected\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP] general treatment strategies include bed rest and supportive treatments , ensuring sufficient energy intake , maintaining a constant internal environment ( water , electrolytes and other internal environment factors ) and monitoring vital signs\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "in vitro experiments showed that ifn α effectively inhibits the replication\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP] ribavirin is a nucleoside analogue with broad antiviral activity . it can prevent the replication of rna and dna viruses by suppressing the activity of inosine monophosphate dehydrogenase , which is required for the synthesis of guanosine triphosphate ( gtp ) . ribavirin was widely used to treat sars patients with or without concomitant use of steroids during the outbreak of sars in hong kong [ 74 ] [ 75 ] [ 76 ] . thus , ribavirin could be considered as a treatment option for covid - 19 patients .\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "suppresses the recently emerged novel cov ( sars - cov - 2 ) in vitro\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "arbidol and arbidol mesylate were shown to have a potent inhibitory effect in reducing the reproduction\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "inhibit sars - cov and mers - cov in vivo\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "cynk - 001 , an allogeneic , off - theshelf , umbilical cord blood - derived nk cell therapy , to the treatment of the newly emerged sars - cov - 2 infection\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "convalescent plasma could , theoretically , be a promising option for the treatment and prevention of sars - cov - 2 infection\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "remission of the sars - cov - 2 epidemic may depend on the development of monoclonal antibodies\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3207\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "it has been confirmed that sars - cov - 2 enters lung cells by binding to ace2\n",
      "3545\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3545\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "3981\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "initially\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "the pathogen of covid - 19 has been attributed to severe acute respiratory syndrome coronavirus 2\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "treatment of covid - 19 has been mostly restricted to supportive measures\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "hydroxychloroquine and chloroquine are traditional anti - malarial and autoimmune disease drugs . they have been shown to control the sars - cov - 2 infection in vitro 61 , 68\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "acei / arb treatment potentially induces overexpression of ace2 , which subsequently increases the susceptibility of host cells to sars - cov - 2 invasion and the risk of infection or worsening the severity of disease\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "dilated horse anti - sars - cov - 2 serum could cross - neutralize sars - cov - 2\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP] general discomforts , typically fever , fatigue , headache , muscle ache , diarrhea , etc . signs and symptoms of lung and airway abnormalities\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "real - time pcr test positive for sars - cov - 2 . viral sequencing showing high homogeneity to sars - cov - 2 . serum test positive for sars - cov - 2 specific igm and igg .\n",
      "4000\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "viral\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "confirmed cases with the sars - cov - 2 were identified as a positive result of a high - throughput sequencing or an rt - pcr assay for respiratory specimens including nasal and pharyngeal swab specimens , bronchoalveolar lavage fluid , sputum , or bronchial aspirates or a positive result of anti - sars - cov - 2 igm / igg or the titer of anti - sars - cov - 2 igg antibody in the recovery period was four times or more higher than in the acute period\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4098\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4099\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "diagnosing sars - cov - 2 with antibodies [SEP] two trained coders completed the coding process independently - one researcher coded the entire sample while a second coder coded 17 random selected articles\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "bed bugs framed in news coverage\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[CLS] diagnosing sars - cov - 2 with antibodies [SEP]\n",
      "4476\n",
      "Diagnosing SARS-COV-2 with antibodies\n",
      "[SEP]\n"
     ]
    }
   ],
   "source": [
    "question = questions[2]\n",
    "for dic in paragraph_list:\n",
    "    for key in dic:\n",
    "        if key == 'index':\n",
    "            continue\n",
    "        ans = answer_question(question, dic[key])\n",
    "        print (dic['index'])\n",
    "        print (question)\n",
    "        print (ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "pytorch-gpu.1-4.m46",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-4:m46"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
